File: pkg/runtime/runtimes/go/pulserpc/diff.go
package pulserpc

import (
	"fmt"
)

func DiffIDL(clientIDL, serverIDL interface{}) []ContractDelta {
	var deltas []ContractDelta

	clientInterfaces := extractInterfaces(clientIDL)
	serverInterfaces := extractInterfaces(serverIDL)

	deltas = append(deltas, diffInterfaces(clientInterfaces, serverInterfaces)...)

	clientStructs := extractStructs(clientIDL)
	serverStructs := extractStructs(serverIDL)
	deltas = append(deltas, diffStructs(clientStructs, serverStructs)...)

	clientEnums := extractEnums(clientIDL)
	serverEnums := extractEnums(serverIDL)
	deltas = append(deltas, diffEnums(clientEnums, serverEnums)...)

	clientErrors := extractErrors(clientIDL)
	serverErrors := extractErrors(serverIDL)
	deltas = append(deltas, diffErrors(clientErrors, serverErrors)...)

	return deltas
}

type idlInterfaces map[string]interface{}
type idlStructs map[string]interface{}
type idlEnums map[string]interface{}
type idlErrors map[string]interface{}

func extractInterfaces(idl interface{}) idlInterfaces {
	result := make(idlInterfaces)
	if dict, ok := idl.(map[string]interface{}); ok {
		if interfaces, ok := dict["interfaces"].([]interface{}); ok {
			for _, ifaceData := range interfaces {
				if ifaceDict, ok := ifaceData.(map[string]interface{}); ok {
					if name, ok := ifaceDict["name"].(string); ok {
						result[name] = ifaceDict
					}
				}
			}
		}
	}
	return result
}

func extractStructs(idl interface{}) idlStructs {
	result := make(idlStructs)
	if dict, ok := idl.(map[string]interface{}); ok {
		if structs, ok := dict["structs"].([]interface{}); ok {
			for _, structData := range structs {
				if structDict, ok := structData.(map[string]interface{}); ok {
					if name, ok := structDict["name"].(string); ok {
						result[name] = structDict
					}
				}
			}
		}
	}
	return result
}

func extractEnums(idl interface{}) idlEnums {
	result := make(idlEnums)
	if dict, ok := idl.(map[string]interface{}); ok {
		if enums, ok := dict["enums"].([]interface{}); ok {
			for _, enumData := range enums {
				if enumDict, ok := enumData.(map[string]interface{}); ok {
					if name, ok := enumDict["name"].(string); ok {
						result[name] = enumDict
					}
				}
			}
		}
	}
	return result
}

func extractErrors(idl interface{}) idlErrors {
	result := make(idlErrors)
	if dict, ok := idl.(map[string]interface{}); ok {
		if errors, ok := dict["errors"].([]interface{}); ok {
			for _, errData := range errors {
				if errDict, ok := errData.(map[string]interface{}); ok {
					if name, ok := errDict["name"].(string); ok {
						result[name] = errDict
					}
				}
			}
		}
	}
	return result
}

func diffInterfaces(client, server idlInterfaces) []ContractDelta {
	var deltas []ContractDelta

	for name, clientIface := range client {
		if serverIface, ok := server[name]; ok {
			deltas = append(deltas, diffInterfaceMethods(name, clientIface, serverIface)...)
		} else {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityInterface,
				EntityName:  name,
				ChangeType:  ChangeRemoved,
				Direction:   DirectionClientHasMore,
				Severity:    ClassifySeverity(EntityInterface, ChangeRemoved, DirectionClientHasMore),
				Description: fmt.Sprintf("Interface '%s' exists in client but not in server", name),
			})
		}
	}

	for name := range server {
		if _, ok := client[name]; !ok {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityInterface,
				EntityName:  name,
				ChangeType:  ChangeAdded,
				Direction:   DirectionClientHasLess,
				Severity:    ClassifySeverity(EntityInterface, ChangeAdded, DirectionClientHasLess),
				Description: fmt.Sprintf("Interface '%s' exists in server but not in client", name),
			})
		}
	}

	return deltas
}

func diffInterfaceMethods(ifaceName string, clientIface, serverIface interface{}) []ContractDelta {
	var deltas []ContractDelta

	clientMethods := extractMethods(clientIface)
	serverMethods := extractMethods(serverIface)

	for name, clientMethod := range clientMethods {
		if serverMethod, ok := serverMethods[name]; ok {
			if !methodsEqual(clientMethod, serverMethod) {
				deltas = append(deltas, ContractDelta{
					EntityType:  EntityMethod,
					EntityName:  ifaceName,
					MemberName:  name,
					ChangeType:  ChangeModified,
					Direction:   DirectionMismatch,
					Severity:    ClassifySeverity(EntityMethod, ChangeModified, DirectionMismatch),
					Description: fmt.Sprintf("Method '%s' in interface '%s' has mismatched signatures", name, ifaceName),
				})
			}
		} else {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityMethod,
				EntityName:  ifaceName,
				MemberName:  name,
				ChangeType:  ChangeRemoved,
				Direction:   DirectionClientHasMore,
				Severity:    ClassifySeverity(EntityMethod, ChangeRemoved, DirectionClientHasMore),
				Description: fmt.Sprintf("Method '%s' in interface '%s' exists in client but not in server", name, ifaceName),
			})
		}
	}

	for name := range serverMethods {
		if _, ok := clientMethods[name]; !ok {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityMethod,
				EntityName:  ifaceName,
				MemberName:  name,
				ChangeType:  ChangeAdded,
				Direction:   DirectionClientHasLess,
				Severity:    ClassifySeverity(EntityMethod, ChangeAdded, DirectionClientHasLess),
				Description: fmt.Sprintf("Method '%s' in interface '%s' exists in server but not in client", name, ifaceName),
			})
		}
	}

	return deltas
}

func extractMethods(iface interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	if dict, ok := iface.(map[string]interface{}); ok {
		if methods, ok := dict["methods"].([]interface{}); ok {
			for _, method := range methods {
				if methodDict, ok := method.(map[string]interface{}); ok {
					if name, ok := methodDict["name"].(string); ok {
						result[name] = methodDict
					}
				}
			}
		}
	}
	return result
}

func methodsEqual(a, b interface{}) bool {
	aMap, aOk := a.(map[string]interface{})
	bMap, bOk := b.(map[string]interface{})
	if !aOk || !bOk {
		return a == b
	}

	aParams := aMap["parameters"]
	bParams := bMap["parameters"]
	if !mapsEqual(aParams, bParams) {
		return false
	}

	aReturn := aMap["returnType"]
	bReturn := bMap["returnType"]
	if !mapsEqual(aReturn, bReturn) {
		return false
	}

	return true
}

func mapsEqual(a, b interface{}) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	aMap, aOk := a.(map[string]interface{})
	bMap, bOk := b.(map[string]interface{})
	if !aOk || !bOk {
		return interfaceEqual(a, b)
	}
	if len(aMap) != len(bMap) {
		return false
	}
	for k, v := range aMap {
		if bv, ok := bMap[k]; !ok || !mapsEqual(v, bv) {
			return false
		}
	}
	return true
}

func interfaceEqual(a, b interface{}) bool {
	switch a.(type) {
	case []interface{}:
		bSlice, ok := b.([]interface{})
		if !ok {
			return false
		}
		aSlice := a.([]interface{})
		if len(aSlice) != len(bSlice) {
			return false
		}
		for i := range aSlice {
			if !interfaceEqual(aSlice[i], bSlice[i]) {
				return false
			}
		}
		return true
	case map[string]interface{}:
		return mapsEqual(a, b)
	case string, float64, int, bool:
		return a == b
	default:
		return a == b
	}
}

func diffStructs(client, server idlStructs) []ContractDelta {
	var deltas []ContractDelta

	for name, clientStruct := range client {
		if serverStruct, ok := server[name]; ok {
			deltas = append(deltas, diffStructFields(name, clientStruct, serverStruct)...)
		} else {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityStruct,
				EntityName:  name,
				ChangeType:  ChangeRemoved,
				Direction:   DirectionClientHasMore,
				Severity:    ClassifySeverity(EntityStruct, ChangeRemoved, DirectionClientHasMore),
				Description: fmt.Sprintf("Struct '%s' exists in client but not in server", name),
			})
		}
	}

	for name := range server {
		if _, ok := client[name]; !ok {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityStruct,
				EntityName:  name,
				ChangeType:  ChangeAdded,
				Direction:   DirectionClientHasLess,
				Severity:    ClassifySeverity(EntityStruct, ChangeAdded, DirectionClientHasLess),
				Description: fmt.Sprintf("Struct '%s' exists in server but not in client", name),
			})
		}
	}

	return deltas
}

func diffStructFields(structName string, clientStruct, serverStruct interface{}) []ContractDelta {
	var deltas []ContractDelta

	clientFields := extractFields(clientStruct)
	serverFields := extractFields(serverStruct)

	for name, clientField := range clientFields {
		if serverField, ok := serverFields[name]; ok {
			typeChanged, optionalityChanged, wasRequired, isRequired := fieldsEqualDetailed(clientField, serverField)
			if typeChanged {
				deltas = append(deltas, ContractDelta{
					EntityType:  EntityField,
					EntityName:  structName,
					MemberName:  name,
					ChangeType:  ChangeModified,
					Direction:   DirectionMismatch,
					Severity:    ClassifySeverity(EntityField, ChangeModified, DirectionMismatch),
					Description: fmt.Sprintf("Field '%s' in struct '%s' has changed type", name, structName),
				})
			} else if optionalityChanged {
				if wasRequired && !isRequired {
					deltas = append(deltas, ContractDelta{
						EntityType:  EntityField,
						EntityName:  structName,
						MemberName:  name,
						ChangeType:  ChangeModified,
						Direction:   DirectionClientHasLess,
						Severity:    ClassifySeverity(EntityField, ChangeModified, DirectionClientHasLess, "made_optional"),
						Description: fmt.Sprintf("Field '%s' in struct '%s' optionality changed from required to optional", name, structName),
					})
				} else if !wasRequired && isRequired {
					deltas = append(deltas, ContractDelta{
						EntityType:  EntityField,
						EntityName:  structName,
						MemberName:  name,
						ChangeType:  ChangeModified,
						Direction:   DirectionClientHasLess,
						Severity:    ClassifySeverity(EntityField, ChangeModified, DirectionClientHasLess, "made_required"),
						Description: fmt.Sprintf("Field '%s' in struct '%s' optionality changed from optional to required", name, structName),
					})
				}
			}
		} else {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityField,
				EntityName:  structName,
				MemberName:  name,
				ChangeType:  ChangeRemoved,
				Direction:   DirectionClientHasMore,
				Severity:    ClassifySeverity(EntityField, ChangeRemoved, DirectionClientHasMore),
				Description: fmt.Sprintf("Field '%s' in struct '%s' exists in client but not in server", name, structName),
			})
		}
	}

	for name, serverField := range serverFields {
		if _, ok := clientFields[name]; !ok {
			isRequired := false
			if opt, ok := serverField.(map[string]interface{}); ok {
				if required, ok := opt["optional"].(bool); ok && !required {
					isRequired = true
				}
			}
			severity := ClassifySeverity(EntityField, ChangeAdded, DirectionClientHasLess, func() string {
				if isRequired {
					return "required"
				}
				return "optional"
			}())
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityField,
				EntityName:  structName,
				MemberName:  name,
				ChangeType:  ChangeAdded,
				Direction:   DirectionClientHasLess,
				Severity:    severity,
				Description: fmt.Sprintf("Field '%s' in struct '%s' exists in server but not in client", name, structName),
			})
		}
	}

	return deltas
}

func extractFields(structData interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	if dict, ok := structData.(map[string]interface{}); ok {
		if fields, ok := dict["fields"].([]interface{}); ok {
			for _, field := range fields {
				if fieldMap, ok := field.(map[string]interface{}); ok {
					if name, ok := fieldMap["name"].(string); ok {
						result[name] = fieldMap
					}
				}
			}
		}
	}
	return result
}

func fieldsEqual(a, b interface{}) bool {
	typeChanged, _, _, _ := fieldsEqualDetailed(a, b)
	return !typeChanged
}

func fieldsEqualDetailed(a, b interface{}) (typeChanged bool, optionalityChanged bool, wasRequired bool, isRequired bool) {
	if a == nil && b == nil {
		return false, false, false, false
	}
	aMap, aOk := a.(map[string]interface{})
	bMap, bOk := b.(map[string]interface{})
	if !aOk || !bOk {
		return a != b, false, false, false
	}
	aType := aMap["type"]
	bType := bMap["type"]
	typeChanged = !mapsEqual(aType, bType)

	aOptional := getFieldOptional(aMap)
	bOptional := getFieldOptional(bMap)
	wasRequired = !aOptional
	isRequired = !bOptional
	optionalityChanged = aOptional != bOptional

	return typeChanged, optionalityChanged, wasRequired, isRequired
}

func getFieldOptional(fieldMap map[string]interface{}) bool {
	if opt, ok := fieldMap["optional"].(bool); ok {
		return opt
	}
	return false
}

func diffEnums(client, server idlEnums) []ContractDelta {
	var deltas []ContractDelta

	for name, clientEnum := range client {
		if serverEnum, ok := server[name]; ok {
			deltas = append(deltas, diffEnumValues(name, clientEnum, serverEnum)...)
		} else {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityEnum,
				EntityName:  name,
				ChangeType:  ChangeRemoved,
				Direction:   DirectionClientHasMore,
				Severity:    ClassifySeverity(EntityEnum, ChangeRemoved, DirectionClientHasMore),
				Description: fmt.Sprintf("Enum '%s' exists in client but not in server", name),
			})
		}
	}

	for name := range server {
		if _, ok := client[name]; !ok {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityEnum,
				EntityName:  name,
				ChangeType:  ChangeAdded,
				Direction:   DirectionClientHasLess,
				Severity:    ClassifySeverity(EntityEnum, ChangeAdded, DirectionClientHasLess),
				Description: fmt.Sprintf("Enum '%s' exists in server but not in client", name),
			})
		}
	}

	return deltas
}

func diffEnumValues(enumName string, clientEnum, serverEnum interface{}) []ContractDelta {
	var deltas []ContractDelta

	clientValues := extractEnumValues(clientEnum)
	serverValues := extractEnumValues(serverEnum)

	for name := range clientValues {
		if _, ok := serverValues[name]; !ok {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityEnum,
				EntityName:  enumName,
				MemberName:  name,
				ChangeType:  ChangeRemoved,
				Direction:   DirectionClientHasMore,
				Severity:    ClassifySeverity(EntityEnum, ChangeRemoved, DirectionClientHasMore),
				Description: fmt.Sprintf("Enum value '%s' in enum '%s' exists in client but not in server", name, enumName),
			})
		}
	}

	for name := range serverValues {
		if _, ok := clientValues[name]; !ok {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityEnum,
				EntityName:  enumName,
				MemberName:  name,
				ChangeType:  ChangeAdded,
				Direction:   DirectionClientHasLess,
				Severity:    ClassifySeverity(EntityEnum, ChangeAdded, DirectionClientHasLess),
				Description: fmt.Sprintf("Enum value '%s' in enum '%s' exists in server but not in client", name, enumName),
			})
		}
	}

	return deltas
}

func extractEnumValues(enumData interface{}) map[string]bool {
	result := make(map[string]bool)
	if dict, ok := enumData.(map[string]interface{}); ok {
		if values, ok := dict["values"].([]interface{}); ok {
			for _, value := range values {
				if valueMap, ok := value.(map[string]interface{}); ok {
					if name, ok := valueMap["name"].(string); ok {
						result[name] = true
					}
				}
			}
		}
	}
	return result
}

func diffErrors(client, server idlErrors) []ContractDelta {
	var deltas []ContractDelta

	for name := range client {
		if _, ok := server[name]; !ok {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityError,
				EntityName:  name,
				ChangeType:  ChangeRemoved,
				Direction:   DirectionClientHasMore,
				Severity:    ClassifySeverity(EntityError, ChangeRemoved, DirectionClientHasMore),
				Description: fmt.Sprintf("Error '%s' exists in client but not in server", name),
			})
		}
	}

	for name := range server {
		if _, ok := client[name]; !ok {
			deltas = append(deltas, ContractDelta{
				EntityType:  EntityError,
				EntityName:  name,
				ChangeType:  ChangeAdded,
				Direction:   DirectionClientHasLess,
				Severity:    ClassifySeverity(EntityError, ChangeAdded, DirectionClientHasLess),
				Description: fmt.Sprintf("Error '%s' exists in server but not in client", name),
			})
		}
	}

	return deltas
}

File: pkg/runtime/runtimes/python/pulserpc/diff.py
"""IDL diff functionality for contract verification"""

from typing import Any, Dict, List, Tuple
from .types import (
    ContractDelta, EntityType, ChangeType, Direction, Severity
)


def diff_idl(client_idl: Dict[str, Any], server_idl: Dict[str, Any]) -> List[ContractDelta]:
    """Compute a directional structural diff between client and server IDL"""
    deltas: List[ContractDelta] = []

    client_interfaces = _extract_interfaces(client_idl)
    server_interfaces = _extract_interfaces(server_idl)
    deltas.extend(_diff_interfaces(client_interfaces, server_interfaces))

    client_structs = _extract_structs(client_idl)
    server_structs = _extract_structs(server_idl)
    deltas.extend(_diff_structs(client_structs, server_structs))

    client_enums = _extract_enums(client_idl)
    server_enums = _extract_enums(server_idl)
    deltas.extend(_diff_enums(client_enums, server_enums))

    client_errors = _extract_errors(client_idl)
    server_errors = _extract_errors(server_idl)
    deltas.extend(_diff_errors(client_errors, server_errors))

    return deltas


def _extract_interfaces(idl: Dict[str, Any]) -> Dict[str, Any]:
    result = {}
    for iface_data in idl.get('interfaces', []):
        name = iface_data.get('name')
        if name:
            result[name] = iface_data
    return result


def _extract_structs(idl: Dict[str, Any]) -> Dict[str, Any]:
    result = {}
    for struct_data in idl.get('structs', []):
        name = struct_data.get('name')
        if name:
            result[name] = struct_data
    return result


def _extract_enums(idl: Dict[str, Any]) -> Dict[str, Any]:
    result = {}
    for enum_data in idl.get('enums', []):
        name = enum_data.get('name')
        if name:
            result[name] = enum_data
    return result


def _extract_errors(idl: Dict[str, Any]) -> Dict[str, Any]:
    result = {}
    for err_data in idl.get('errors', []):
        name = err_data.get('name')
        if name:
            result[name] = err_data
    return result


def _diff_interfaces(client: Dict[str, Any], server: Dict[str, Any]) -> List[ContractDelta]:
    deltas: List[ContractDelta] = []

    for name, client_iface in client.items():
        if name in server:
            deltas.extend(_diff_interface_methods(name, client_iface, server[name]))
        else:
            deltas.append(ContractDelta(
                entity_type=EntityType.INTERFACE,
                entity_name=name,
                member_name='',
                change_type=ChangeType.REMOVED,
                direction=Direction.CLIENT_HAS_MORE,
                severity=Severity.ERROR,
                description=f"Interface '{name}' exists in client but not in server"
            ))

    for name in server:
        if name not in client:
            deltas.append(ContractDelta(
                entity_type=EntityType.INTERFACE,
                entity_name=name,
                member_name='',
                change_type=ChangeType.ADDED,
                direction=Direction.CLIENT_HAS_LESS,
                severity=Severity.INFO,
                description=f"Interface '{name}' exists in server but not in client"
            ))

    return deltas


def _diff_interface_methods(iface_name: str, client_iface: Dict[str, Any], server_iface: Dict[str, Any]) -> List[ContractDelta]:
    deltas: List[ContractDelta] = []
    client_methods = _extract_methods(client_iface)
    server_methods = _extract_methods(server_iface)

    for name, client_method in client_methods.items():
        if name in server_methods:
            if not _methods_equal(client_method, server_methods[name]):
                deltas.append(ContractDelta(
                    entity_type=EntityType.METHOD,
                    entity_name=iface_name,
                    member_name=name,
                    change_type=ChangeType.MODIFIED,
                    direction=Direction.MISMATCH,
                    severity=Severity.ERROR,
                    description=f"Method '{name}' in interface '{iface_name}' has mismatched signatures"
                ))
        else:
            deltas.append(ContractDelta(
                entity_type=EntityType.METHOD,
                entity_name=iface_name,
                member_name=name,
                change_type=ChangeType.REMOVED,
                direction=Direction.CLIENT_HAS_MORE,
                severity=Severity.ERROR,
                description=f"Method '{name}' in interface '{iface_name}' exists in client but not in server"
            ))

    for name in server_methods:
        if name not in client_methods:
            deltas.append(ContractDelta(
                entity_type=EntityType.METHOD,
                entity_name=iface_name,
                member_name=name,
                change_type=ChangeType.ADDED,
                direction=Direction.CLIENT_HAS_LESS,
                severity=Severity.WARNING,
                description=f"Method '{name}' in interface '{iface_name}' exists in server but not in client"
            ))

    return deltas


def _extract_methods(iface: Dict[str, Any]) -> Dict[str, Any]:
    result = {}
    for method in iface.get('methods', []):
        name = method.get('name')
        if name:
            result[name] = method
    return result


def _methods_equal(a: Dict[str, Any], b: Dict[str, Any]) -> bool:
    if not _maps_equal(a.get('parameters'), b.get('parameters')):
        return False
    if not _maps_equal(a.get('returnType'), b.get('returnType')):
        return False
    return True


def _maps_equal(a: Any, b: Any) -> bool:
    if a is None and b is None:
        return True
    if a is None or b is None:
        return False
    if isinstance(a, dict) and isinstance(b, dict):
        if len(a) != len(b):
            return False
        for k, v in a.items():
            if k not in b or not _maps_equal(v, b[k]):
                return False
        return True
    if isinstance(a, list) and isinstance(b, list):
        if len(a) != len(b):
            return False
        for i in range(len(a)):
            if not _maps_equal(a[i], b[i]):
                return False
        return True
    return a == b


def _diff_structs(client: Dict[str, Any], server: Dict[str, Any]) -> List[ContractDelta]:
    deltas: List[ContractDelta] = []

    for name, client_struct in client.items():
        if name in server:
            deltas.extend(_diff_struct_fields(name, client_struct, server[name]))
        else:
            deltas.append(ContractDelta(
                entity_type=EntityType.STRUCT,
                entity_name=name,
                member_name='',
                change_type=ChangeType.REMOVED,
                direction=Direction.CLIENT_HAS_MORE,
                severity=Severity.ERROR,
                description=f"Struct '{name}' exists in client but not in server"
            ))

    for name in server:
        if name not in client:
            deltas.append(ContractDelta(
                entity_type=EntityType.STRUCT,
                entity_name=name,
                member_name='',
                change_type=ChangeType.ADDED,
                direction=Direction.CLIENT_HAS_LESS,
                severity=Severity.INFO,
                description=f"Struct '{name}' exists in server but not in client"
            ))

    return deltas


def _diff_struct_fields(struct_name: str, client_struct: Dict[str, Any], server_struct: Dict[str, Any]) -> List[ContractDelta]:
    deltas: List[ContractDelta] = []
    client_fields = _extract_fields(client_struct)
    server_fields = _extract_fields(server_struct)

    for name, client_field in client_fields.items():
        if name in server_fields:
            type_changed, optionality_changed, was_required, is_required = _fields_equal_detailed(client_field, server_fields[name])
            if type_changed:
                deltas.append(ContractDelta(
                    entity_type=EntityType.FIELD,
                    entity_name=struct_name,
                    member_name=name,
                    change_type=ChangeType.MODIFIED,
                    direction=Direction.MISMATCH,
                    severity=Severity.ERROR,
                    description=f"Field '{name}' in struct '{struct_name}' has changed type"
                ))
            elif optionality_changed:
                if was_required and not is_required:
                    deltas.append(ContractDelta(
                        entity_type=EntityType.FIELD,
                        entity_name=struct_name,
                        member_name=name,
                        change_type=ChangeType.MODIFIED,
                        direction=Direction.CLIENT_HAS_LESS,
                        severity=Severity.INFO,
                        description=f"Field '{name}' in struct '{struct_name}' optionality changed from required to optional"
                    ))
                elif not was_required and is_required:
                    deltas.append(ContractDelta(
                        entity_type=EntityType.FIELD,
                        entity_name=struct_name,
                        member_name=name,
                        change_type=ChangeType.MODIFIED,
                        direction=Direction.CLIENT_HAS_LESS,
                        severity=Severity.WARNING,
                        description=f"Field '{name}' in struct '{struct_name}' optionality changed from optional to required"
                    ))
        else:
            deltas.append(ContractDelta(
                entity_type=EntityType.FIELD,
                entity_name=struct_name,
                member_name=name,
                change_type=ChangeType.REMOVED,
                direction=Direction.CLIENT_HAS_MORE,
                severity=Severity.INFO,
                description=f"Field '{name}' in struct '{struct_name}' exists in client but not in server"
            ))

    for name, server_field in server_fields.items():
        if name not in client_fields:
            is_required = _is_field_required(server_field)
            severity = _classify_severity(EntityType.FIELD, ChangeType.ADDED, Direction.CLIENT_HAS_LESS, "required" if is_required else "optional")
            deltas.append(ContractDelta(
                entity_type=EntityType.FIELD,
                entity_name=struct_name,
                member_name=name,
                change_type=ChangeType.ADDED,
                direction=Direction.CLIENT_HAS_LESS,
                severity=severity,
                description=f"Field '{name}' in struct '{struct_name}' exists in server but not in client"
            ))

    return deltas


def _extract_fields(struct_data: Dict[str, Any]) -> Dict[str, Any]:
    result = {}
    for field in struct_data.get('fields', []):
        name = field.get('name')
        if name:
            result[name] = field
    return result


def _fields_equal_detailed(a: Dict[str, Any], b: Dict[str, Any]) -> Tuple[bool, bool, bool, bool]:
    type_changed = not _maps_equal(a.get('type'), b.get('type'))
    a_optional = _get_field_optional(a)
    b_optional = _get_field_optional(b)
    was_required = not a_optional
    is_required = not b_optional
    optionality_changed = a_optional != b_optional
    return type_changed, optionality_changed, was_required, is_required


def _get_field_optional(field: Dict[str, Any]) -> bool:
    return bool(field.get('optional', False))


def _is_field_required(field: Dict[str, Any]) -> bool:
    return not field.get('optional', False)


def _diff_enums(client: Dict[str, Any], server: Dict[str, Any]) -> List[ContractDelta]:
    deltas: List[ContractDelta] = []

    for name, client_enum in client.items():
        if name in server:
            deltas.extend(_diff_enum_values(name, client_enum, server[name]))
        else:
            deltas.append(ContractDelta(
                entity_type=EntityType.ENUM,
                entity_name=name,
                member_name='',
                change_type=ChangeType.REMOVED,
                direction=Direction.CLIENT_HAS_MORE,
                severity=Severity.WARNING,
                description=f"Enum '{name}' exists in client but not in server"
            ))

    for name in server:
        if name not in client:
            deltas.append(ContractDelta(
                entity_type=EntityType.ENUM,
                entity_name=name,
                member_name='',
                change_type=ChangeType.ADDED,
                direction=Direction.CLIENT_HAS_LESS,
                severity=Severity.WARNING,
                description=f"Enum '{name}' exists in server but not in client"
            ))

    return deltas


def _diff_enum_values(enum_name: str, client_enum: Dict[str, Any], server_enum: Dict[str, Any]) -> List[ContractDelta]:
    deltas: List[ContractDelta] = []
    client_values = _extract_enum_values(client_enum)
    server_values = _extract_enum_values(server_enum)

    for name in client_values:
        if name not in server_values:
            deltas.append(ContractDelta(
                entity_type=EntityType.ENUM,
                entity_name=enum_name,
                member_name=name,
                change_type=ChangeType.REMOVED,
                direction=Direction.CLIENT_HAS_MORE,
                severity=Severity.WARNING,
                description=f"Enum value '{name}' in enum '{enum_name}' exists in client but not in server"
            ))

    for name in server_values:
        if name not in client_values:
            deltas.append(ContractDelta(
                entity_type=EntityType.ENUM,
                entity_name=enum_name,
                member_name=name,
                change_type=ChangeType.ADDED,
                direction=Direction.CLIENT_HAS_LESS,
                severity=Severity.WARNING,
                description=f"Enum value '{name}' in enum '{enum_name}' exists in server but not in client"
            ))

    return deltas


def _extract_enum_values(enum_data: Dict[str, Any]) -> Dict[str, bool]:
    result = {}
    for value in enum_data.get('values', []):
        name = value.get('name')
        if name:
            result[name] = True
    return result


def _diff_errors(client: Dict[str, Any], server: Dict[str, Any]) -> List[ContractDelta]:
    deltas: List[ContractDelta] = []

    for name in client:
        if name not in server:
            deltas.append(ContractDelta(
                entity_type=EntityType.ERROR,
                entity_name=name,
                member_name='',
                change_type=ChangeType.REMOVED,
                direction=Direction.CLIENT_HAS_MORE,
                severity=Severity.INFO,
                description=f"Error '{name}' exists in client but not in server"
            ))

    for name in server:
        if name not in client:
            deltas.append(ContractDelta(
                entity_type=EntityType.ERROR,
                entity_name=name,
                member_name='',
                change_type=ChangeType.ADDED,
                direction=Direction.CLIENT_HAS_LESS,
                severity=Severity.INFO,
                description=f"Error '{name}' exists in server but not in client"
            ))

    return deltas


def _classify_severity(entity_type: EntityType, change_type: ChangeType, direction: Direction, extra: str = '') -> Severity:
    if entity_type == EntityType.STRUCT:
        if change_type == ChangeType.REMOVED and direction == Direction.CLIENT_HAS_MORE:
            return Severity.ERROR
        if change_type == ChangeType.ADDED and direction == Direction.CLIENT_HAS_LESS:
            return Severity.INFO

    elif entity_type == EntityType.FIELD:
        if change_type == ChangeType.MODIFIED and direction == Direction.MISMATCH:
            return Severity.ERROR
        if change_type == ChangeType.REMOVED and direction == Direction.CLIENT_HAS_MORE:
            return Severity.INFO
        if change_type == ChangeType.ADDED and direction == Direction.CLIENT_HAS_LESS:
            if extra == 'required':
                return Severity.ERROR
            return Severity.INFO
        if change_type == ChangeType.MODIFIED and direction == Direction.CLIENT_HAS_LESS:
            if extra == 'made_required':
                return Severity.WARNING
            if extra == 'made_optional':
                return Severity.INFO
            return Severity.INFO

    elif entity_type == EntityType.METHOD:
        if change_type == ChangeType.REMOVED and direction == Direction.CLIENT_HAS_MORE:
            return Severity.ERROR
        if change_type == ChangeType.ADDED and direction == Direction.CLIENT_HAS_LESS:
            return Severity.WARNING
        if change_type == ChangeType.MODIFIED and direction == Direction.MISMATCH:
            return Severity.ERROR

    elif entity_type == EntityType.ENUM:
        if change_type == ChangeType.REMOVED and direction == Direction.CLIENT_HAS_MORE:
            return Severity.WARNING
        if change_type == ChangeType.ADDED and direction == Direction.CLIENT_HAS_LESS:
            return Severity.WARNING

    elif entity_type == EntityType.ERROR:
        if change_type == ChangeType.REMOVED and direction == Direction.CLIENT_HAS_MORE:
            return Severity.INFO
        if change_type == ChangeType.ADDED and direction == Direction.CLIENT_HAS_LESS:
            return Severity.INFO

    elif entity_type == EntityType.INTERFACE:
        if change_type == ChangeType.REMOVED and direction == Direction.CLIENT_HAS_MORE:
            return Severity.ERROR
        if change_type == ChangeType.ADDED and direction == Direction.CLIENT_HAS_LESS:
            return Severity.INFO

    return Severity.INFO